### PR TITLE
Expand JSDoc template generation locations

### LINKED ISSUE #4447: JSDoc auto-fill only works on functions
Currently JSDoc auto-fill for comment scaffolding seems to only work for functions.  This should also work for other forms like classes, methods, namespaces, etc

--------------

*Edit by @DanielRosenwasser*
Adapted from list below:

- [x] methods
- [x] classes
- [x] class constructors
- [x] namespaces
- [x] variables
- [ ] property assignments
- [ ] property declarations
- [ ] computed properties
- [ ] enums
- [ ] enum members
- [ ] index signatures
- [ ] interfaces
- [ ] type aliases
- [ ] IIFEs?
- [ ] parenthesized expressions

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This pull request addresses portions of #4447.

We currently now cover
*  methods
* classes
* class constructors
* namespaces
* variables

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -7025,16 +7025,35 @@ namespace ts {
             }
 
             // TODO: add support for:
-            // - methods
-            // - constructors
-            // - class decls
-            let containingFunction = <FunctionDeclaration>getAncestor(tokenAtPos, SyntaxKind.FunctionDeclaration);
+            // - enums/enum members
```
Comment (yuit) [inline src/services/services.ts]: update comment about valid position preceding a function declaration.

File: src/services/services.ts [hunk 2]
```diff
@@ -7025,16 +7025,35 @@ namespace ts {
...
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.Constructor:
+                    case SyntaxKind.ClassDeclaration:
+                    case SyntaxKind.VariableStatement:
+                        break findOwner;
+                    case SyntaxKind.SourceFile:
+                        return undefined;
+                    case SyntaxKind.ModuleDeclaration:
+                        // We don't want to give back a JSDoc for the 'b' in 'module a.b'.
+                        if (commentOwner.parent.kind === SyntaxKind.ModuleDeclaration) {
```
Comment (yuit) [inline src/services/services.ts]: you don't want to have jsdoc for 'a' as well right?

Comment (DanielRosenwasser) [inline src/services/services.ts]: No, if we hit one of the nested namespace declarations, then the position given is *within* the span of dotted names.

Check out `tests/cases/fourslash/docCommentTemplateNamespacesAndModules01.ts`

File: src/services/services.ts [hunk 3]
```diff
@@ -7043,9 +7062,15 @@ namespace ts {
             // TODO: call a helper method instead once PR #4133 gets merged in.
             const newLine = host.getNewLine ? host.getNewLine() : "\r\n";
 
-            let docParams = parameters.reduce((prev, cur, index) =>
-                prev +
-                indentationStr + " * @param " + (cur.name.kind === SyntaxKind.Identifier ? (<Identifier>cur.name).text : "param" + index) + newLine, "");
+            let docParams = "";
```
Comment (yuit) [inline src/services/services.ts]: Should we try add type annotation in to the @param ?

**Comment by yuit** [comment]
Could you add tests for these cases?
```ts
function foo(...a) {}
```

```ts
function foo({a,b,c}){}
```

**Comment by yuit** [comment]
:cookie:

**Pull request merged**